src/kOS.Safe/Communication/GenericMessageQueue.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Collections.Generic;
+using kOS.Safe.Serialization;
+using kOS.Safe.Encapsulation;
+using kOS.Safe.Exceptions;
+using System.Linq;
+using kOS.Safe.Encapsulation.Suffixes;
+using kOS.Safe;
+
+namespace kOS.Safe.Communication
+{
+    public class GenericMessageQueue<M> : IDumper where M : BaseMessage
+    {
+        /// <summary>
+        /// This stores a mapping: moment in time -> list of messages that arrive at that time.
+        /// We need the queue to be constructed this way because SortedList doesn't allow for duplicate keys.
+        /// </summary>
+        private SortedList<double, List<M>> queue = new SortedList<double, List<M>>();
+        private CurrentTimeProvider timeProvider;
+
+        private List<M> Messages {
+            get {
+                return queue.Aggregate(new List<M>(), (acc, l) => { acc.AddRange(l.Value); return acc; } );
+            }
+        }
+
+        private List<M> ReceivedMessages {
+            get {
+                return Messages.Where(IsReceived).ToList();
+            }
+        }
```

Comment:
(See elsewhere in the PR comments for the comment with the word "curly brace" in big letters.)